### DISCUSSION TITLE
Custom data type with `strip_whitespace` in Pydantic V2

### DISCUSSION THREAD

**Discussion by prryplatypus** [pydantic/pydantic#6469] [Question] [answered]
Hello. I'm trying to migrate to Pydantic V2, but I'm a bit stuck with how to create a reusable datatype that is a strict type and with which whitespaces are stripped. Back with Pydantic V1, this is what I used:

```py
from pydantic import StrictStr

class Username(StrictStr):
    min_length = 3
    max_length = 20
    strip_whitespace = True
```

I'm looking for the equivalent of this in Pydantic V2. For now this option works at runtime, but Mypy isn't happy with it.

```py
from pydantic import constr

Username = constr(
    strip_whitespace=True, min_length=3, max_length=20, strict=True
)
```

This is the only alternative I've found for my code to be happy when typechecking and at runtime, but it looks nasty...
```py
from typing import TYPE_CHECKING
from pydantic import constr


Username = (
    str
    if TYPE_CHECKING
    else constr(
        strip_whitespace=True, min_length=3, max_length=20, strict=True
    )
)
```

Would there be a better way to convert this?

Many thanks in advance for your help.

**Comment by adriangb** [comment, marked as answer]
Try something like this:

```py
from typing_extensions import Annotated

from pydantic import Field, AfterValidator

Username = Annotated[str, AfterValidator(lambda x: str.strip(x)), Field(min_length=3, max_length=20)]

from pydantic import TypeAdapter, ValidationError

ta = TypeAdapter(Username)

print(ta.validate_python('adriangb' + ' ' * 10))
try:
    ta.validate_python('a' * 100 + ' ' * 10)
except ValidationError as exc:
   print(exc)
else:
    assert False
```

You might need to play around with the order of the length constraints and the whitespace stripping.

**Reply by prryplatypus** [reply]
Hi, thanks for your help.

This example fails (expectedly) with a string such as "  f  " when using the type adapter since the final string "f" is shorter than 3 charscters. However, when using the field in a model it doesn't seem to error, but I believe it should (I've tried swapping the stripping and length constraint too). Am I missing something?

```py
from typing_extensions import Annotated
from pydantic import BaseModel, Field, AfterValidator

Username = Annotated[str, AfterValidator(lambda x: str.strip(x)), Field(min_length=3, max_length=20)]

class MyModel(BaseModel):
	username: Username

MyModel(username="  f  ")  # This should raise a validation error because the final string "f" is shorter than `min_length`, but doesn't actually raise anything
```

**Reply by adriangb** [reply]
That’s what I meant about playing around with the order of things. If you want a specific order, you might have to go little bit more custom than that. Or it might be as easy as using a `BeforeValidator`

**Reply by prryplatypus** [reply]
Oh alright, I misunderstood what you meant with that. This feels so much harder now than with V1 😭.
Thanks for your help in pointing me in the correct direction 🙁

**Reply by adriangb** [reply]
I'm sorry it's more difficult. I'm not sure why we removed `str_strip_whitespace` but it seems like a reasonable feature request to add. This case is particularly tricky because the `min_length` constraint gets applied before after coercion but before the Python string is created, and hence before the `AfterValidator`. So there is really no way to insert custom logic between these two steps. Since there's no real coercion for strings you can use a `BeforeValidator` to get it to work:

```python
from typing_extensions import Annotated
from pydantic import BaseModel, Field, BeforeValidator

Username = Annotated[str, BeforeValidator(lambda x: str.strip(str(x))), Field(min_length=3, max_length=20)]

class MyModel(BaseModel):
	username: Username

MyModel(username="  f  ")
```

But I admit that's a lot of mental gymnastics to jump through for something that was pretty simple in v1, the easiest thing to do is probably to just add back `str_strip_whitespace` to `Field` or create some sort of `ConStr()` marker for annotated (`Annotated[str, ConStr(strip_whitespace=True), Field(max_length=10, ...)]` or something like that)

**Reply by prryplatypus** [reply]
Created a feature request: https://github.com/pydantic/pydantic/issues/6531
